File: app/src/main/java/com/w9jds/marketbot/ui/SearchActivity.java
package com.w9jds.marketbot.ui;

import android.app.SearchManager;
import android.content.Intent;
import android.support.design.widget.Snackbar;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.view.View;
import android.widget.ProgressBar;

import com.w9jds.marketbot.R;
import com.w9jds.marketbot.classes.models.MarketItemBase;
import com.w9jds.marketbot.data.loader.GroupsLoader;
import com.w9jds.marketbot.ui.adapters.MarketGroupsAdapter;
import com.w9jds.marketbot.data.DataLoadingSubject;

import java.util.List;

import butterknife.Bind;
import butterknife.ButterKnife;

public class SearchActivity extends AppCompatActivity implements DataLoadingSubject.DataLoadingCallbacks {

    @Bind(R.id.market_groups) RecyclerView recyclerView;
    @Bind(R.id.dataloading_progress) ProgressBar progressBar;
    @Bind(R.id.toolbar) Toolbar toolbar;

    private ActionBar actionBar;
    private GroupsLoader searchLoader;
    private MarketGroupsAdapter adapter;
    private LinearLayoutManager layoutManager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_search);
        ButterKnife.bind(this);

        setSupportActionBar(toolbar);
        actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayShowHomeEnabled(true);
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setHomeButtonEnabled(true);
        }

        layoutManager = new LinearLayoutManager(this);
        recyclerView.setLayoutManager(layoutManager);
        recyclerView.setItemAnimator(new DefaultItemAnimator());

        adapter = new MarketGroupsAdapter(this);
        recyclerView.setAdapter(adapter);

        handleIntent(getIntent());
    }

    @Override
    protected void onNewIntent(Intent intent) {
        handleIntent(intent);
    }

    private void createDataManager() {

        searchLoader = new GroupsLoader(this) {
            @Override
            public void onProgressUpdate(int page, int totalPages, String message) {

            }

            @Override
            public void onDataLoaded(List<? extends MarketItemBase> data) {
                adapter.updateCollection(data);
            }
        };

        searchLoader.registerLoadingCallback(this);
    }

    private void handleIntent(Intent intent) {

        if (searchLoader == null) {
            createDataManager();
        }

        if (Intent.ACTION_SEARCH.equals(intent.getAction())) {
            String query = intent.getStringExtra(SearchManager.QUERY);

            actionBar.setTitle(query);
            searchLoader.searchMarketTypes(query);
        }
    }

    @Override
    public void dataStartedLoading() {
        progressBar.setVisibility(View.VISIBLE);
    }

    @Override
    public void dataFinishedLoading() {
        progressBar.setVisibility(View.GONE);
    }

    @Override
    public void dataFailedLoading(String errorMessage) {
//        Snackbar.make(baseView, errorMessage, Snackbar.LENGTH_LONG).show();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                onBackPressed();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

}

File: app/src/main/java/com/w9jds/marketbot/ui/SearchActivity.kt
package com.w9jds.marketbot.ui

import android.app.SearchManager
import android.content.Intent
import android.databinding.DataBindingUtil

import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.LinearLayoutManager
import android.view.MenuItem
import com.google.firebase.database.DataSnapshot
import com.w9jds.marketbot.R
import com.w9jds.marketbot.data.DataLoadingSubject
import com.w9jds.marketbot.data.loader.SearchLoader
import com.w9jds.marketbot.databinding.ActivitySearchBinding
import com.w9jds.marketbot.ui.adapters.GroupsAdapter
import kotlinx.android.synthetic.main.app_bar.view.*

class SearchActivity: AppCompatActivity(), DataLoadingSubject.DataLoadingCallbacks {

    private lateinit var binding: ActivitySearchBinding
    private lateinit var adapter: GroupsAdapter
    private lateinit var loader: SearchLoader

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_search)

        if (binding.baseView.toolbar != null) {
            setSupportActionBar(binding.baseView.toolbar)
        }

        if (supportActionBar != null) {
            supportActionBar?.setDisplayHomeAsUpEnabled(true)
        }

        loader = object: SearchLoader(this) {
            override fun onDataLoaded(data: List<DataSnapshot>) {
                adapter.updateItems(data)
            }
        }

        binding.resultsList.layoutManager = LinearLayoutManager(this)
        binding.resultsList.itemAnimator = DefaultItemAnimator()

        adapter = GroupsAdapter(this, null)
        binding.resultsList.adapter = adapter

        handleIntent(intent)
    }

    override fun onNewIntent(intent: Intent?) {
        if (intent != null) {
            handleIntent(intent)
        }
    }

    override fun dataStartedLoading() {
        binding.swipeRefresh.isRefreshing = true
    }

    override fun dataFinishedLoading() {
        binding.swipeRefresh.isRefreshing = false
    }

    override fun dataFailedLoading(errorMessage: String) {
        binding.swipeRefresh.isRefreshing = false
        Snackbar.make(binding.baseView, errorMessage, Snackbar.LENGTH_LONG).show()
    }

    private fun handleIntent(intent: Intent) {
        if (Intent.ACTION_SEARCH == intent.action) {
            val query: String = intent.getStringExtra(SearchManager.QUERY)
            supportActionBar?.title = query
            loader.loadSearchResults(query)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        return when (item?.itemId) {
            R.id.home -> {
                onBackPressed()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }


}